File: frontend/src/utils/promptSanitizer.ts
/**
 * Redact secret material from agent-instruction prompts before they are
 * handed to a third-party LLM provider.
 *
 * The 2026-04-15 code review (critical finding #1) flagged that
 * `InAppAgentPanel` was POSTing the raw ``instructions`` markdown
 * emitted by ``/test-plans/{id}/execute`` and
 * ``/test-plans/generate-recon`` directly to the configured LLM
 * provider.  Those instruction blocks include:
 *
 *   - The one-time ``X-API-Key: nm_agent_...`` header (so the agent
 *     can authenticate as the user to BlueStick's own API).
 *   - Inlined scanner integration credentials (Nessus/OpenVAS/Nuclei/Burp
 *     access + secret keys from ``integration_service.decrypt_integration``).
 *
 * Piping these through an LLM turns every configured provider into a
 * potential credential sink — the provider logs the request body, the
 * operator can see it, and any prompt caching at the provider retains
 * it indefinitely.
 *
 * The fix is to sanitize on the client side before the request ever
 * leaves the browser.  The full (unredacted) instructions are still
 * available for the user to copy into a terminal-side agent (Claude
 * Code, Codex, etc) — only the in-app LLM path gets the redacted
 * version.  The redaction is non-reversible and the placeholder text
 * tells the agent explicitly that secrets were stripped and must be
 * requested from the operator out-of-band.
 */

const REDACTED = '[REDACTED — request from operator out-of-band]';

/**
 * Strip secrets from an agent-instruction prompt.
 *
 * Patterns removed:
 *  1. ``X-API-Key: nm_agent_...`` — the BlueStick agent API token
 *     the user was just handed.  The LLM doesn't need it; approval
 *     still happens in the user's terminal / in-app UI, and the
 *     coordinator model forbids autonomous tool execution anyway.
 *  2. Markdown bullet lines that inline scanner credentials emitted
 *     by ``_integration_block`` in ``agent_prompt_service.py``:
 *     lines matching ``  - Access key:``, ``  - Secret key:``,
 *     ``  - Password:``, ``  - Username:``, ``  - API key:``,
 *     ``  - PDCP token:``, ``  - Secret:``.  The label stays, the
 *     value becomes REDACTED.
 *  3. Bare ``nm_agent_xxx`` tokens anywhere in the text (defense in
 *     depth against the key appearing outside the header block).
 *
 * The function is intentionally conservative: if a caller's prompt
 * contains content that looks like a key pattern, it gets redacted.
 * False positives are cheap (the LLM sees ``[REDACTED]`` instead of a
 * non-secret string); false negatives are expensive (the key leaks).
 */
export function sanitizePromptForLlm(prompt: string): string {
  if (!prompt) return prompt;
  let out = prompt;

  // 1. Strip the X-API-Key line.  Match the whole line so the fence
  //    formatting stays intact.
  out = out.replace(
    /^\s*X-API-Key:\s*nm_agent_[A-Za-z0-9_-]+\s*$/gm,
    `X-API-Key: ${REDACTED}`,
  );

  // 2. Strip inlined integration credential bullets.  The prompt
  //    template uses ``  - <Label>: `<value>``` markdown, so we match
  //    the label and replace everything after the colon on that line.
  const credLabels = [
    'Access key',
    'Secret key',
    'Password',
    'Username',
    'API key',
    'PDCP token',
    'Secret',
  ];
  for (const label of credLabels) {
    const re = new RegExp(
      String.raw`^(\s*-\s*${label}:\s*)\`[^\`]*\`\s*$`,
      'gim',
    );
    out = out.replace(re, `$1\`${REDACTED}\``);
  }

  // 3. Defense-in-depth: catch any bare agent-key tokens elsewhere in
  //    the text (e.g. if a future template references the key outside
  //    the header block).  Width matches the X-API-Key matcher above
  //    (``+``, not ``{20,}``) so a short-suffix token inlined in prose
  //    is caught by the backstop too — the ``nm_agent_`` prefix plus a
  //    word boundary is distinctive enough to avoid false positives.
  out = out.replace(/nm_agent_[A-Za-z0-9_-]+/g, REDACTED);

  // 4. Well-known third-party secret formats.  These prompts are
  //    forwarded to the configured LLM provider, so a pasted cloud / API
  //    credential (not just a BlueStick agent key) would land in that
  //    provider's request log.  Keep in sync with the backend
  //    app/services/prompt_sanitizer.py pattern list.
  out = out.replace(/\bAKIA[0-9A-Z]{16}\b/g, REDACTED); // AWS access key id
  out = out.replace(/\bsk-[A-Za-z0-9_-]{20,}\b/g, REDACTED); // OpenAI-style secret key
  out = out.replace(/\bgh[pousr]_[A-Za-z0-9]{20,}\b/g, REDACTED); // GitHub tokens
  out = out.replace(/\bxox[baprs]-[A-Za-z0-9-]{10,}\b/g, REDACTED); // Slack tokens
  out = out.replace(/\bAIza[0-9A-Za-z_-]{35}\b/g, REDACTED); // Google API key
  // JSON Web Token (three base64url segments)
  out = out.replace(
    /\beyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\b/g,
    REDACTED,
  );
  // Bearer auth tokens — keep the scheme, redact the credential
  out = out.replace(/Bearer\s+[A-Za-z0-9._-]{20,}/gi, `Bearer ${REDACTED}`);

  return out;
}

File: backend/app/services/prompt_sanitizer.py
"""
Server-side prompt sanitizer.

Audit finding H2: the client-side ``promptSanitizer.ts`` stripped
agent API keys and inlined scanner credentials before the
``InAppAgentPanel`` posted to ``/llm-providers/{id}/complete``, but
the backend endpoint itself did no scrubbing.  A user who bypassed
the frontend (direct curl, scripted client, compromised analyst
session) could deliberately include a real agent API key or
credential in the ``prompt`` body and the backend would forward the
full text to the configured LLM provider, where it would land in the
provider's request log.

This module mirrors ``frontend/src/utils/promptSanitizer.ts`` at the
bullet / pattern level.  Keep them in sync when patterns change —
the frontend version is defense-in-depth (makes the intent visible
to users and prevents accidental leakage via the happy path), and
the server version is the enforcement point (catches anything that
bypasses the frontend).

The sanitizer is intentionally aggressive: false positives (redacting
non-secret text that happens to match a pattern) are cheap — the LLM
sees ``[REDACTED]`` instead of a benign string.  False negatives (a
real secret slipping through) are expensive, so we err on the side
of over-stripping.
"""

from __future__ import annotations

import re
from typing import Optional


_REDACTED = "[REDACTED — out of band]"

# Patterns are applied in order.  Each entry is (compiled regex,
# replacement string).  The replacement uses Python's ``re.sub``
# semantics so ``\1`` / ``\g<name>`` work for capturing groups.
_PATTERNS: list[tuple[re.Pattern, str]] = [
    # 1. X-API-Key header line with an nm_agent_* token.  Matches
    #    the whole line so the markdown/code-fence shape stays
    #    intact.
    (
        re.compile(
            r"X-API-Key:\s*nm_agent_[A-Za-z0-9_-]+",
            re.IGNORECASE,
        ),
        f"X-API-Key: {_REDACTED}",
    ),
    # 2. Inlined scanner-credential bullets emitted by
    #    ``agent_prompt_service._integration_block``.  The template
    #    shape is ``  - <Label>: `<value>``` (backtick-wrapped
    #    value after a label).  Redact the value, keep the label so
    #    the LLM can still reason about the structure of the prompt.
    (
        re.compile(
            r"^(\s*-\s*(?:Access key|Secret key|Password|Username|API key|PDCP token|Secret)\s*:\s*)`[^`]*`",
            re.IGNORECASE | re.MULTILINE,
        ),
        rf"\1`{_REDACTED}`",
    ),
    # 3. Defense in depth — any bare ``nm_agent_...`` token elsewhere
    #    in the text gets stripped.  The 20+ char minimum avoids
    #    false-matching shorter strings that happen to start with
    #    ``nm_agent_`` (prompt-version strings, etc).
    (
        re.compile(r"nm_agent_[A-Za-z0-9_-]{20,}"),
        _REDACTED,
    ),
    # 4. Well-known third-party secret formats.  These prompts are
    #    forwarded to whatever LLM provider the operator configured, so
    #    a pasted cloud / API credential (not just a BlueStick agent key)
    #    would land in that provider's request log.  Same cheap-redaction
    #    posture: err on the side of over-stripping.  Keep in sync with
    #    frontend/src/utils/promptSanitizer.ts.
    (re.compile(r"\bAKIA[0-9A-Z]{16}\b"), _REDACTED),            # AWS access key id
    (re.compile(r"\bsk-[A-Za-z0-9_-]{20,}\b"), _REDACTED),       # OpenAI-style secret key
    (re.compile(r"\bgh[pousr]_[A-Za-z0-9]{20,}\b"), _REDACTED),  # GitHub tokens
    (re.compile(r"\bxox[baprs]-[A-Za-z0-9-]{10,}\b"), _REDACTED),  # Slack tokens
    (re.compile(r"\bAIza[0-9A-Za-z_-]{35}\b"), _REDACTED),       # Google API key
    (  # JSON Web Token (three base64url segments)
        re.compile(r"\beyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\b"),
        _REDACTED,
    ),
    (  # Bearer auth tokens — keep the scheme, redact the credential
        re.compile(r"(?i)\bBearer\s+[A-Za-z0-9._\-]{20,}"),
        f"Bearer {_REDACTED}",
    ),
]


def sanitize_for_llm(text: Optional[str]) -> Optional[str]:
    """Return ``text`` with sensitive patterns replaced.

    ``None`` and empty strings pass through unchanged so callers can
    uniformly call ``sanitize_for_llm(body.system)`` even when
    ``system`` is optional.  Every other input gets the full pattern
    sweep.
    """
    if not text:
        return text
    out = text
    for pattern, replacement in _PATTERNS:
        out = pattern.sub(replacement, out)
    return out
